include/grpc/grpc_security.h

```diff
@@ -249,19 +249,40 @@ typedef struct {
   void *reserved;
 } grpc_auth_metadata_context;
 
+/** Maximum number of credentials returnable by a credentials plugin via
+    a synchronous return. */
+#define GRPC_METADATA_CREDENTIALS_PLUGIN_SYNC_MAX 4
```

Comment:
Can we make it configurable? Php need to get the callback always invoked on the same thread.

Reply:
No, we can't make it configurable.  For performance reasons, we need to allocate space for the return parameters on the stack instead of heap-allocating it, so the number has to be set at compile time.

This limitation may go away in the future with some metadata API changes that @ctiller has in mind.  But for now, if PHP needs to have a plugin return more then 4 metadata elements, then it needs to use the async API.